File: src/app/clients/chatgptClient.js
const fetch = (...args) => import('node-fetch').then(({default: fetch}) => fetch(...args));
const { logInfo, logError } = require('../../logger');
const { config } = require('../../config');

async function callChatGPT(query) {
  try {
    const response = await fetch(config.apis.openai.apiUrl, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${config.apis.openai.apiKey}`
      },
      body: JSON.stringify({
        model: config.apis.openai.model,
        messages: [
          {
            role: 'user',
            content: query
          }
        ],
        max_tokens: config.apis.openai.maxTokens,
        temperature: config.apis.openai.temperature
      })
    });

    if (!response.ok) {
      throw new Error(`OpenAI API error: ${response.status} ${response.statusText}`);
    }

    const data = await response.json();
    logInfo('ChatGPT response received', { usage: data.usage });
    
    return data.choices[0]?.message?.content || 'No response from ChatGPT.';
  } catch (err) {
    logError('ChatGPT API error', err);
    return 'Error connecting to ChatGPT. Please try again later.';
  }
}

module.exports = { callChatGPT };

File: python_app/app/clients/chatgpt_client.py
from __future__ import annotations

from typing import Any, Callable, Dict, List, Optional

import requests

from ...config import config
from ...logger import log_error, log_info


def call_chatgpt(query: str) -> str:
    if not config.openai.api_key:
        log_error("ChatGPT API error", ValueError("OPENAI_API_KEY is not configured"))
        return "ChatGPT integration is not configured."

    prefers_responses_api = "responses" in config.openai.api_url.rstrip("/").split("/")[-1]
    try:
        return _call_openai_api(
            url=config.openai.api_url,
            payload=_build_payload(query, prefers_responses_api),
            parser=_parse_response if prefers_responses_api else _parse_chat_completion,
        )
    except requests.HTTPError as http_error:
        response = http_error.response
        body = None
        status = None
        try:
            if response is not None:
                status = response.status_code
                body = response.text
        except Exception:
            body = None
        log_error(
            "OpenAI Responses call failed",
            {
                "status": status,
                "body": body,
            },
        )
        return "Live answer unavailable (Responses API failed)."
    except requests.RequestException as error:
        log_error("ChatGPT API error", error)
        return "Error connecting to ChatGPT. Please try again later."


def _call_openai_api(
    *,
    url: str,
    payload: Dict[str, Any],
    parser: Callable[[Dict[str, Any]], Optional[str]],
) -> str:
    headers = {
        "Content-Type": "application/json",
        "Authorization": f"Bearer {config.openai.api_key}",
    }

    response = requests.post(
        url,
        headers=headers,
        json=payload,
        timeout=30,
    )
    response.raise_for_status()
    try:
        data: Dict[str, Any] = response.json()
    except ValueError as error:
        log_error("OpenAI API returned invalid JSON", error)
        return "OpenAI API returned an unexpected response."

    log_info("ChatGPT response received", {"usage": data.get("usage")})
    content = parser(data)
    return content or "No response from ChatGPT."


def _build_payload(query: str, use_responses: bool) -> Dict[str, Any]:
    if use_responses:
        payload = {
            "model": config.openai.model,
            "input": [
                {
                    "role": "user",
                    "content": [{"type": "input_text", "text": query}],
                }
            ],
            "max_output_tokens": config.openai.max_tokens,
            "temperature": config.openai.temperature,
        }
        if config.openai.browser_enabled:
            tool_type = "web_search_preview_2025_03_11"
            payload["tools"] = [{"type": tool_type}]
            payload["tool_choice"] = "required"
        return payload

    return {
        "model": config.openai.model,
        "messages": [{"role": "user", "content": query}],
        "max_tokens": config.openai.max_tokens,
        "temperature": config.openai.temperature,
    }


def _parse_response(data: Dict[str, Any]) -> Optional[str]:
    output: List[Dict[str, Any]] = data.get("output") or []
    for item in output:
        content_entries: List[Dict[str, Any]] = item.get("content") or []
        for entry in content_entries:
            if entry.get("type") in {"text", "output_text"}:
                text_value: Optional[str] = entry.get("text")
                if text_value:
                    return text_value
    return None


def _parse_chat_completion(data: Dict[str, Any]) -> Optional[str]:
    choices: List[Dict[str, Any]] = data.get("choices") or []
    if choices:
        message = choices[0].get("message", {})
        content = message.get("content")
        if content:
            return str(content)
    return None
